src/solveh.rs

```diff
@@ -254,20 +254,51 @@ where
 /// An interface for calculating determinants of Hermitian (or real symmetric) matrix refs.
 pub trait DeterminantH {
     type Output;
+    type SignLnOutput;
```

Comment:
```rust
pub trait DeterminantH {
  type Output;
  type Sign;
  fn deth(&self) -> Self::Output;
  fn sign_ln_deth(&self) -> (Self::Sign, Self::Output);
}
```

is more natural. You may define `SignLnOutput` because its implementation for `Array2`. This `Result<_>` has been introduced in https://github.com/termoshtt/ndarray-linalg/pull/87, but it can be dropped I think. What do you think?

Reply:
Yeah, I don't like the `SignLnOutput` associated type either. I added it because because the `DeterminantH` impl for `ArrayBase<S, Ix2>` needs to return a result (because the factorization can fail), while the impl for `BKFactorized<S>` doesn't need to return a result (since the factorization is already done). The return types are

```rust
impl<A, S> DeterminantH for BKFactorized<S> {
// ...
    fn deth(&self) -> A::Real {}
    fn sign_ln_deth(&self) -> (A::Real, A::Real) {}
}

impl<A, S> DeterminantH for ArrayBase<S, Ix2>
// ...
{
    fn deth(&self) -> Result<A::Real> {}
    fn sign_ln_deth(&self) -> Result<(A::Real, A::Real)> {}
}
```

I suppose it would be possible to get rid of the extra associated type if `ArrayBase<S, Ix2>::sign_ln_deth()` returned `(Result<A::Real>, Result<A::Real>)` by cloning the error in the `Err` case, but that's less convenient for the user and a little weird because it would return two copies of the error.

A good option is to add methods on `BKFactorized<S>` that don't return a `Result`, and change `DeterminantH` to always return a `Result`. When you call `factorized.deth()`, the methods would be used (which is great), but you could still pass a `BKFactorized` to a function with a `T: DeterminantH` bound. In other words:

```rust
impl<A, S> BKFactorized<S>
where
    A: Scalar,
    S: Data<Elem = A>,
{
    pub fn deth(&self) -> A::Real {
        // ...
    }
    pub fn sign_ln_deth(&self) -> (A::Real, A::Real) {
        // ...
    }
}

pub trait DeterminantH<A: Scalar> {
    fn deth(&self) -> Result<A::Real>;
    fn sign_ln_deth(&self) -> Result<(A::Real, A::Real)>;
}

impl<A, S> DeterminantH<A> for BKFactorized<S>
where
    A: Scalar,
    S: Data<Elem = A>,
{
    fn deth(&self) -> Result<A::Real> {
        Ok(self.deth())
    }
    fn sign_ln_deth(&self) -> Result<(A::Real, A::Real)> {
        Ok(self.sign_ln_deth())
    }
}

impl<A, S> DeterminantH<A> for ArrayBase<S, Ix2>
where
    A: Scalar,
    S: Data<Elem = A>,
{
    fn deth(&self) -> Result<A::Real> {
        // ...
    }
    fn sign_ln_deth(&self) -> Result<(A::Real, A::Real)> {
        // ...
    }
}
```

What do you think?

Reply:
> A good option is to add methods on BKFactorized<S> that don't return a Result, and change DeterminantH to always return a Result

This should be better, but it needs bumping up version. 

> because the factorization can fail

This should be my mistake not to consider when the factorization can fail. We have to investigate the case where the factorization fail, and must return `Result<_>`.

To keep this PR simple, I accept your original one using `SignLnOutput`. I will create another issue about the failure of factorization.

Reply:
> This should be better, but it needs bumping up version.

I think adding the `*ln_det*()` methods and `SignLnOutput` type is technically a breaking change too (since the user could have implemented the `Determinant*` traits for their own types).

> To keep this PR simple, I accept your original one using `SignLnOutput`.

Okay, that's fine. After this PR is merged, I'll submit another PR following my proposal above.